### ISSUE TITLE
Does the heat generation of solar panels scale with how much energy they're providing?

### ISSUE THREAD

**Issue by TheUnfetteredOne** [endless-sky/endless-sky#2508]
That is, will a solar panel produce more heat closer to the system center than when farther away?
If this isn't the case right now, should it be?

**Comment by Elyssaen** [comment]
Doesn't look like it. They just define `heat generation` like any other outfit, so it's static.

Maybe it should work as you say. Would have to define a new attribute, `solar heat generation`, that was added to your ship's heat in the same place, but scaled by the `scale` variable that's used to scale `solar collection`.

If there's some consensus that it should work that way, I'll code and PR it.

**Comment by tehhowch** [comment]
It is kinda silly that we are told a planet is too hot to land on, but not too hot to orbit indefinitely :)

:+1: for heat generation from based on your distance to the system star. 
:-1: for possibly needing to rebalance all fleets because of extra "heat damage" to all ships involved.

`solar heat generation` should consider the star sprite, btw - those old red stars aren't nearly as hot.
One possibility would be to scale based on the system star and then on your ships mass - as in general most ships are made of similar materials with similar densities and the densities of typical things such as weapons, bunks, reactors, etc does not vary from ship to ship, so the mass of your ship scales with its volume, and the solar fluence scales with its surface area.

so something like
`solar heat gen = star_energy_output (J/s) * distance scaling (1/m^2) * exposed_surface_area (m^2)`
If we assume ships to be cylinders with aspect ratio H = 4R, then R is determined from the ship's average density and its baseAttributes.Mass value (since storing outfits/cargo inside doesn't alter the volume):
`R= ( (Mass/avg_density )/(4 * pi) )^1/3`
which lets you determine the exposed surface area (assuming 1/2 is shaded from the star):
` esa = 0.5 * 10 * pi * R^2 ` -> ` esa ~ 2.91 * (Mass/ avg_dens)^(2/3)`
Given the average density of most metals is 2k-10k  kg/m^3  and the spaceship needs at least some non-metal sections, you could probably peg the average ship density to somewhere in the range of 1-4k  kg/m^3, and introduce a density scaling factor as a function of total outfit space, ship type, etc. 
This leaves you with solar heat generation dropping off with 1/r2 (or the ramscoop efficiency profile, probably), and increasing with Mass^(2/3). 

Since maximum allowed heat scales linearly with Mass (as Mass()*100), solar heat generation will affect fighters and light craft more than heavy ships (which are already able to tailor their cooling more effectively than lighter craft). On the plus side, this means Sestor swarms might hurt themselves if you are strategic about how to engage them, and people will stop complaining about how OP fighters are :laughing:

**Comment by Elyssaen** [comment]
That's... a _lot_ more modelling there. I thought what this issue was talking about was simply scaling the heat generation of solar collectors parallel to how the solar collection itself scales. (With a balancing check / halving the total, as I think that scale factor can hit 2.)

I wasn't suggesting distance to a star generate additional heat for all ships. It's an interesting suggestion, but... another topic. A long one, probably! I do like the idea that different stars might provide faster/slower scooping and more/less solar collection energy. About at the same level of complexity would be using distance to individual stars, not to the system centre.

But the rest of your post... you've gone and run a bit ahead of @TheUnfetteredOne's issue there. : )

**Comment by jafdy** [comment]
If it's "too hot to land on" then really it means it is "too hot for you to walk around on" but that sounds silly. Also think Venus. 

Red giants may not be as hot on the surface as other stars (including the sun) but in space only the radiation power output matters as there is no conduction or convection. Luminosity is the single thing that will affect stellar heating on your craft. The same would be true for solar power generation if the conversion efficiency is equal across all bands.

In practice modeling this as anything more interesting then as a constant multiplier from solar power generation is likely to be a waste of time.

**Comment by tehhowch** [comment]
I don't think it would be too difficult to get a per-ship multiplier based on the ship's defined base attributes (category would set density, & Mass). Once you have that, solar heat generation is just a constant multiplier that scales with distance to the star (with optional star dependence).

I feel it would be nonsensical to define solar heat generation only for ships that have solar panels though

**Comment by Elyssaen** [comment]
@tehhowch, why nonsensical? It's measuring the waste heat produced when converting your 'fuel source' into solar collection. It's parallel to heat generation in a reactor, it just ramps with the solar collector's % load.

It's _not_ heat radiating onto your ship's hull from nearby stars.

**Comment by tehhowch** [comment]
If it's called `solar heat generation`, that's what it must be. 

You could define some property that is associated with resistive heating losses though, as that would make sense for a solar panel pumping out more energy than it was rated for

**Comment by Elyssaen** [comment]
Okay, I get the ambiguity. But getting into resistive heating losses, quantum efficiency or whatever else seems a bit too detailed: every other outfit just says `heat generation`, it doesn't spell out where the inefficiencies occur. I was suggesting `solar heat generation` in parallel to that.

Are there any good alternatives that don't try to go into any more detail about what processes generate the waste heat? Maybe `heat generation (scaled)` or a phrase of that form?

**Comment by tehhowch** [comment]
`waste heat factor` is a good one. It wouldn't even really need to be displayed numerically (like the rampscoop function), just noted in the description of solar panels that if they output extra power, they will also output extra heat.

**Comment by Elyssaen** [comment]
So just to be clear, we're talking about something like:

![waste heat factor](https://cloud.githubusercontent.com/assets/23544894/25896792/80dc4b0c-357d-11e7-8778-05aa6a773af1.JPG)
![solar heat generation](https://cloud.githubusercontent.com/assets/23544894/25896822/9deee9de-357d-11e7-86e6-7b528f7131ba.JPG)

`waste heat factor` would really confuse me, personally.

**Comment by tehhowch** [comment]
This is where what I thought was meant has an advantage over just multiplying the outfit heat generation by some waste heat factor: those outfits were balanced on having a certain heat load. Increasing it isn't likely to have a useful effect. 
Applying some additional heat to all ships based on relative solar system position is a New Thing and doesn't affect an individual outfit's balancing vs its competition - it becomes a new strategy to manipulate in flight

**Comment by EndrosG** [comment]
I have a quote from @endless-sky as he recently changed [CONTRIBUTING.md](https://github.com/endless-sky/endless-sky/blob/master/CONTRIBUTING.md)

> If requesting a new feature, first ask yourself: will this make the game more fun or interesting? Remember that this is a game, not a simulator. Changes will not be made purely for the sake of realism, especially if they introduce needless complexity or aggravation.

**I like the idea that a solar collector generates more heat with higher output (both scaling with distance to the star(s)).** But all beyond, i.e. `stellar heat` independent from any solar collectors etc., is too much in my eyes.
I personally have no idea about the original lore/story on that, if stellar heat is already taken into account when balancing a ships heat dissipation, or if it is meant that ships with shields are protected against external heat sources or or or... But introducing `stellar heat` and already knowing to need to rebalance all ships seems unlikely to get accepted by MZ.

-----------------------------

It would be more interestering (simulation-wise), but still not necessarily more fun (playing-wise), to have ramscoop first of all depend on the number of stars at all.
Right now, it does not matter if we have zero stars, one star, two stars or any other number in a system.

[Ship.cpp, Lines 774+](https://github.com/endless-sky/endless-sky/blob/master/source/Ship.cpp#L774) says:
```
	// Update ship supply levels.
	if(!isDisabled)
	{
		// Ramscoops work much better when close to the system center. Even if a
		// ship has no ramscoop, it can harvest a tiny bit of fuel by flying
		// close to the star.
		double scale = .2 + 1.8 / (.001 * position.Length() + 1);
		fuel += .03 * scale * (sqrt(attributes.Get("ramscoop")) + .05 * scale);
		fuel = min(fuel, attributes.Get("fuel capacity"));
		
		energy += scale * attributes.Get("solar collection");
```

Nothing about the number of stellar objects at all.
Even in the "system" holding my ancient metropolis wreckage the ramscoop works.

**Comment by Amazinite** [comment]
Changing how ramscoop and solar collection work based off of the number and type of stars in a system is really outside of the scope of this issue; I'd suggest you open a new one for that issue and remain on the topic of heat generation's ties with solar collection here.

**Comment by TheUnfetteredOne** [comment]
@tehhowch No thanks.

**Comment by tehhowch** [comment]
On topic: making heat generation from solar panels that already have heat generation scale with the distance to the system center (and/or the number of stars in the system) requires a new attribute, as @Elyssaen pointed out in reply 1. This is so that the heat calculations can perform the appropriate multiplication by the scale factor.

I'm not sure this should happen without some other compensation to balance it, as right now, having low (or no) heat generation is something solar modules were balanced on. To simply (and silently) increase the negatives of an outfit is not going to make the game more fun or interesting.
Any thoughts on how decreasing the heat efficiency of solar outfits can be compensated?

**Comment by Amazinite** [comment]
>I'm not sure this should happen without some other compensation to balance it, as right now, having low (or no) heat generation is something solar modules were balanced on. To simply (and silently) increase the negatives of an outfit is not going to make the game more fun or interesting.

How is it increasing the negatives? 100% solar collection = 100% heat generation, so it would at max generate what it does now and beyond that generate less heat. If anything this would be a buff.

**Comment by Elyssaen** [comment]
Yeah, I've implemented it (on a branch of my fork, no PR yet) as Amazinite just described. That's what my screenshots were based on.

**Comment by EndrosG** [comment]
@Amazinite 

quoting again parts of the same code as above:
		// Ramscoops work much better when close to the system center. Even if a
		// ship has no ramscoop, it can harvest a tiny bit of fuel by flying
		// close to the star.
		double scale = .2 + 1.8 / (.001 * position.Length() + 1);
		energy += scale * attributes.Get("solar collection");

Quoting from the [wiki](https://github.com/endless-sky/endless-sky/wiki/CreatingOutfits):
As of v. 0.9.0, ramscoops are more effective near the system center: the fuel gain is multiplied by `.2 + 1.8 / (distance to center / 1000 + 1)`

The closer you get to the systems center, the closer the fuel gain factor can get to 2.0 = 0.2 + 1.8

At a distances of 1000 the formula returns `0.2 + 1.8 / (1000/1000 + 1) = 0.2 + 1.8 / 2 = 0.2 + 0.9 = 1.1`.
And that increases up to almost 2.0 if distance is approx. zero.
So, the worst what can happen is double heat because of double solar generation.

**Comment by Elyssaen** [comment]
@EndrosG is right of course. I stuck in:

``		heat += .5 * scale * attributes.Get("solar heat generation");``

so it varies from 0.1 to 1, not 0.2 to 2, but otherwise scales in the same way as solar collection and ramscoops.

I think it's more 'fair' (show the worst case scenario for heat), but it does mean it's then inconsistent with the solar collection. Personally, I don't think we should be talking about doing anything to make these collector modules _worse_. If we code it with just `scale` and not `.5 * scale`, it'd make sense to halve the value so it's the same net effect: you only get the previous max at the system centre.

**Comment by Elyssaen** [comment]
@TheUnfetteredOne, I've updated Elyssaen#13 with a few changes:

1. Added a tooltip entry for `solar heat generation` (name is still up for debate).
2. IdleHeat should get the correct amount of solar heat generation for your current position.
3. I've changed it to scale exactly like `solar collection`, meaning the value in the Outfitter Panel is actually half the maximum you can reach. I then halved the face values in coalition outfits.txt so that the max heat before and after is the same. (But since it'll almost always be less than that, it's actually a buff.)

Want a PR of this, or any changes? Anyone else have any thought on different ways of doing this (scaling, balancing, IdleHeat, tooltip wording, etc.)?

**Comment by Elyssaen** [comment]
@TheUnfetteredOne, I'd be ready to PR it but for the question label assigned above (vs. mechanics/enhancement). 

Since it's not a priority for me, I'll leave this one with you. [This link](https://github.com/Elyssaen/endless-sky/pull/new/feature/SolarHeatScaling) will make a PR for it. If you want to PR it and it doesn't say "Able to merge", let me know and I'll resolve any conflict.
